spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/dsl/Stream.java

```diff
@@ -0,0 +1,290 @@
+			this.client = client;
+			Assert.hasLength(name, "Stream name can't be empty");
+			this.name = name;
+		}
+
+		public SourceBuilder source(StreamApplication source) {
+			Assert.notNull(source, "Source application can't be null");
+			return new SourceBuilder(source.type(StreamApplication.StreamType.SOURCE),
+					this);
+		}
+
+		public SourceBuilder source(String name){
+			Assert.isTrue(StringUtils.hasLength(name), "Source application can't be null");
+			StreamApplication source = new StreamApplication(name);
+			return new SourceBuilder(source.type(StreamApplication.StreamType.SOURCE), this);
+		}
+
+		/**
+		 * Creates a Stream bypassing the fluent API and just using the provided
+		 * definition
+		 * @param definiton the Stream definition to use
+		 * @return A {@link Stream} object
+		 */
+		public StreamDefinitionBuilder definition(String definiton) {
+			Assert.hasLength(name, "Stream definition can't be empty");
+			this.definition = definiton;
+			return new StreamDefinitionBuilder(this.name, this.client, this.definition);
+		}
+
+		private Stream create() {
+			return new Stream(this.name, this.applications, this.definition, client)
+					.create();
+		}
+	}
+
+	public static class StreamDefinitionBuilder {
+
+		private String name;
+
+		private DataFlowOperations client;
+
+		private String definition;
+
+		private StreamDefinitionBuilder(String name, DataFlowOperations client,
+				String definition) {
+			this.name = name;
+			this.client = client;
+			this.definition = definition;
+		}
+
+		public Stream create() {
+			return new Stream(this.name, Collections.emptyList(), this.definition, client)
+					.create();
+		}
+	}
+
+	public static class SourceBuilder extends BaseBuilder {
+
+		private SourceBuilder(StreamApplication source, StreamNameBuilder parent) {
+			super(source, parent);
+		}
+
+		public ProcessorBuilder processor(StreamApplication processor) {
+			Assert.notNull(processor, "Processor application can't be null");
+			return new ProcessorBuilder(
+					processor.type(StreamApplication.StreamType.PROCESSOR), this.parent);
+		}
+
+		public ProcessorBuilder processor(String name) {
+			Assert.hasLength(name, "Stream definition can't be empty");
+			StreamApplication processor = new StreamApplication(name);
+			return new ProcessorBuilder(
+					processor.type(StreamApplication.StreamType.PROCESSOR), this.parent);
+		}
+
+		public SinkBuilder sink(StreamApplication sink) {
+			Assert.notNull(sink, "Sink application can't be null");
+			return new SinkBuilder(sink.type(StreamApplication.StreamType.SINK),
+					this.parent);
+		}
+
+		public SinkBuilder sink(String name){
+			Assert.hasLength(name, "Stream definition can't be empty");
+			StreamApplication sink = new StreamApplication(name);
+			return new SinkBuilder(sink.type(StreamApplication.StreamType.SINK),
+					this.parent);
+		}
+
+	}
+
+	public static class ProcessorBuilder extends BaseBuilder {
```

Comment:
We could add some check in the processor builder if the same name is repeated for the application name for the given Stream and suggest using a label instead. This way, we can fail fast on the client side.